### PR TITLE
[homekit] add support for min/max values for temperature

### LINKED ISSUE #3400: [homekit] Make thermostat temperature limits configurable
<a alt="bounty" href="https://www.bountysource.com/issues/56349501"></a>

It would be great if temperature limits for thermostats could be configured independently.
I know there are already `minimumTemperature` and `maximumTemperature` settings, but they have an impact on all temperature items (like outdoor temperature sensors).

To get practical limits for thermostat temperature sliders separate settings would be required.

/cc @beowulfe

### PR DESCRIPTION
**Pull request by yfre**
HomeKit specification supports custom min/max values as well as the min step for some characteristics, including temperature. For instance, an outside temperature sensor can have a different temperature range than an inside temperature sensor. 

Java-HAP library added support for custom min/max values recently. 

This PR add supports for min/max/step to HomeKit binding. 

changes:
- update to latest Java-HAP library (minor update)
- add support for configurable min/max/step values at item level for temperatures. 
- reduce level of one logging warning which is rather common and harmless

Closes #3400

Signed-off-by: Eugen Freiter <freiter@gmx.de>

### CODE REVIEW

File: bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/accessories/AbstractHomekitAccessoryImpl.java [hunk 1]
```diff
@@ -174,6 +174,34 @@ protected void unsubscribe(HomekitCharacteristicType characteristicType) {
...
+     * The type for number is BigDecimal.
+     * @param characteristicType characteristic type
+     * @param key configuration key
+     * @param defaultValue default value
+     * @param <T> expected type
+     * @return configuration value
+     */
+    protected <T> T getAccessoryConfiguration(@NonNull HomekitCharacteristicType characteristicType, @NonNull String key, @NonNull T defaultValue) {
+        final Optional<HomekitTaggedItem> characteristic = getCharacteristic(characteristicType);
+        return characteristic.isPresent()?getItemConfiguration(characteristic.get(), key,defaultValue):defaultValue;
```
[the remark this answers is not in the document]
Comment (yfre) [reply bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/accessories/AbstractHomekitAccessoryImpl.java]: not sure why i missed to run spotless.
fixed. now it should be well formatted.

**Comment by cpmeister** [comment]
Just closing a reopening to see if I can get the DCO error to go away.

**Comment by Hobohome** [comment]
I think there is an issue with the [minValue=10.5, maxValue=27]  function in that it does not accept a negative value. 
If I use a value [minValue=-25, maxValue=30] (for a freezer for example) the item file is rejected with a message ...
Configuration model 'custom.items' has errors, therefore ignoring it: [311,124]: no viable alternative at input '-'

Number Coolroom_Temp "Coolroom [%.1f°C ]" <temperature> (gSQLpersist_everyChange) 
		{ channel="mqtt:topic:pi1_mqtt:Coolroom_Sensor:Coolroom_Temp",homekit = "TemperatureSensor.CurrentTemperature" [minValue=-25, maxValue=30],alexa="TemperatureSensor.temperature" }

**Comment by yfre** [comment]
@Hobohome  thank you for reporting. 
i have discovered this as well. but this is an issue in openHab core (it cannot parse negative number)
issue: https://github.com/openhab/openhab-core/issues/1501
fix: https://github.com/openhab/openhab-core/pull/1517

but the fix will be first available with openhab 3.0. we cannot do much until then.
i will add this to the documentation as "known limitations".

**Comment by Hobohome** [comment]
Hay @yfre  and thanks for the reply. Do you know if it is possible still to over-ride the global default min/max in the Homekit config? I read that adding two custom parameters in PaperUI (maximumTemperature and minimumTemperature) this could be done. 
I have tried this and still get an error when the value returned from the sensor is less than 10 degC. Perhaps I have not done this correctly?

**Comment by yfre** [comment]
@Hobohome you have done all correctly. it is documentation that is not up-to-date. the global defaults are not supported anymore. they were dropped in favour of local, item level, settings. 

but if "-" is not supported at item level, maybe we should add global level back.. or support both, check local, if not there, take global.  let me see if this can be easily done

as of now, there is no way to define "-" for temperatures. sorry for that.

**Comment by Hobohome** [comment]
Thanks @yfre . I think having both global and item level would be great. Allow the item level value (when specified) to override the global values.

**Comment by grzegorz914** [comment]
Hi,

minValue and maxValue for Thermostat and Temperature Sensor don’t work in OH 3.0.
Also negative minValue are not acceptable in config.

item:
`temperature_sensor   "Temperature Sensor [%.1f C]"     {homekit="TemperatureSensor" [minValue=0, maxValue=100] } `

`2020-11-08 09:11:19.336 [WARN ] [istics.impl.base.FloatCharacteristic] - Detected value out of range 4.8. Returning min value instead. Characteristic io.github.hapjava.characteristics.impl.thermostat.CurrentTemperatureCharacteristic@16061a5 `

**Comment by pug304** [comment]
> 
> 
> Hi,
> 
> minValue and maxValue for Thermostat and Temperature Sensor don’t work in OH 3.0.
> Also negative minValue are not acceptable in config.
> 
> item:
> `Number temperature_sensor "Temperature Sensor [%.1f C]" {homekit="TemperatureSensor" [minValue=0, maxValue=100] } `
> 
> `2020-11-08 09:11:19.336 [WARN ] [istics.impl.base.FloatCharacteristic] - Detected value out of range 4.8. Returning min value instead. Characteristic io.github.hapjava.characteristics.impl.thermostat.CurrentTemperatureCharacteristic@16061a5 `

it seems issue stil is existing in most recent version unstable 3.0.0~S2072-1

**Comment by yfre** [comment]
@pug304
hm. it should work. i recently tested it with OH3 using items and using UI. 
  what you see in karaf if you typ

`metadata list `
or
`metadata list <item_name>`

**Comment by pug304** [comment]
Metadata [key=homekit:TemperaturGarage_1_ACTUAL_TEMPERATURE, value=TemperatureSensor.CurrentTemperature, configuration=[]]
Metadata [key=stateDescription:TemperaturGarage_1_ACTUAL_TEMPERATURE, value= , configuration=[step=0.1, min=-30, max=40]]
Metadata [key=semantics:TemperaturGarage_1_ACTUAL_TEMPERATURE, value=Point, configuration=[isPointOf=TemperaturGarage]]

due to out of limits oh reports 10° to homekit, i've tested with min=0 ans different steps

**Comment by yfre** [comment]
can you try to close home app and keep it closed for 10-20 seconds and open again?
or add sensor with different name?

home app request the complete list of accessories only at very beginning or at new start after longer time. once it has it, it only asks for actual values and not for meta data.

**Comment by pug304** [comment]
i will give this a try tomorrow :)

**Comment by grzegorz914** [comment]
Hi,
minValue and maxValue working OK but step not.

**Comment by pug304** [comment]
have doen this steps:
- close all Homekit apps, shut down Homekit relay device
- restart  Homekit on one device, no change - and stopped the app after
- removed one temperature sensor and re-added with another name

- also no change :(

error: 18:17:29.892 [WARN ] [ristics.impl.base.FloatCharacteristic] - Detected value out of range 41.0. Returning max value instead. Characteristic io.github.hapjava.characteristics.impl.thermostat.TargetTemperatureCharacteristic@2853999d

may be the error is because of setting a step value

**Comment by grzegorz914** [comment]
After U set metadata with minValue/maxValue/step stop and start OH with:

```
sudo /bin/systemctl stop openhab.service
sudo /bin/systemctl start openhab.service
```

**Comment by pug304** [comment]
- close all Homekit apps, shut down Homekit relay device
- deleted model, things & items
- sudo /bin/systemctl restart openhab.service
- recreated everything
- set minValue/maxValue/Step
- sudo /bin/systemctl restart openhab.service
- set Homekit metadata
- sudo /bin/systemctl restart openhab.service

without success

**Comment by yfre** [comment]
thank you for trying all this. 
strange it is not working. the logic in the code is pretty simple, probably 10 lines in total. 
can you share you exact configuration for one item, which is not working and how exactly you do the configuration. 
this will help to reproduce and fix the bug

**Comment by grzegorz914** [comment]
@pug304 
I think U set minVaule and maxValue in wrong palace, U need to set this in HomeKit metadata config file not in item config.

**Comment by pug304** [comment]
> 
> 
> @pug304
> I think U set minVaule and maxValue in wrong place, U need to set this in HomeKit metadata config file not in the item config itself.

IMHO this changed in OH3, you have to set the values in the item definition

**Comment by grzegorz914** [comment]
8E6036A4-A82C-460A-8115-8E9D77119BA1
6F2B018F-A068-4AF7-B230-FF55EE14674D

**Comment by pug304** [comment]
> 
> 
> thank you for trying all this.
> strange it is not working. the logic in the code is pretty simple, probably 10 lines in total.
> can you share you exact configuration for one item, which is not working and how exactly you do the configuration.
> this will help to reproduce and fix the bug

for sure, no secret ;)

thing code

--------

UID: homematic:HmIP-STHO-A:3014F711A0001F98A9AA0C5E:0010DA499476AE
label: MyTemperaturGarage
thingTypeUID: homematic:HmIP-STHO-A
configuration:
  HMP_0_CYCLIC_INFO_MSG_DIS_UNCHANGED: 0
  HMP_2_COND_TX_CYCLIC_BELOW: false
  HMP_3_EVENT_DELAY_VALUE: 3
  HMP_3_EVENT_RANDOMTIME_VALUE: 1
  HMP_2_EVENT_RANDOMTIME_UNIT: S
  HMP_2_COND_TX_RISING: false
  HMP_3_COND_TX_FALLING: false
  HMP_0_ENABLE_ROUTING: true
  HMP_2_COND_TX_CYCLIC_ABOVE: false
  HMP_3_EVENT_DELAY_UNIT: S
  HMP_0_LOW_BAT_LIMIT: 2.4
  HMP_1_TEMPERATURE_OFFSET: 0
  HMP_2_COND_TX_FALLING: false
  HMP_3_COND_TX_CYCLIC_ABOVE: false
  HMP_2_EVENT_RANDOMTIME_VALUE: 1
  HMP_3_COND_TX_DECISION_ABOVE: 200
  HMP_0_CYCLIC_INFO_MSG_OVERDUE_THRESHOLD: 2
  HMP_2_EVENT_DELAY_UNIT: S
  HMP_2_COND_TX_DECISION_BELOW: 0
  HMP_2_COND_TX_THRESHOLD_LO: 26
  HMP_2_EVENT_DELAY_VALUE: 3
  HMP_0_CYCLIC_INFO_MSG_DIS: 0
  HMP_2_COND_TX_THRESHOLD_HI: 27
  HMP_0_LOCAL_RESET_DISABLED: false
  HMP_0_ARR_TIMEOUT: 10
  HMP_3_COND_TX_THRESHOLD_LO: 70
  HMP_0_DUTYCYCLE_LIMIT: 180
  HMP_3_COND_TX_THRESHOLD_HI: 80
  HMP_3_COND_TX_DECISION_BELOW: 0
  HMP_3_EVENT_RANDOMTIME_UNIT: S
  HMP_3_COND_TX_CYCLIC_BELOW: false
  HMP_2_COND_TX_DECISION_ABOVE: 200
  HMP_0_CYCLIC_INFO_MSG: 1
  HMP_3_COND_TX_RISING: false
bridgeUID: homematic:bridge:3014F711A0001F98A9AA0C5E


------------

Item/Link

grafik

-------------

Model

grafik

-------------

medadata

grafik

grafik

TemperatureSensor tag is set in regarding model 

other then in OH2 (all definitions inside .thing ans .item files)  i've created all in the UI

**Comment by pug304** [comment]
> 
> 
> 8E6036A4-A82C-460A-8115-8E9D77119BA1
> 6F2B018F-A068-4AF7-B230-FF55EE14674D

ah i see....

**Comment by grzegorz914** [comment]
..... but Step not working

**Comment by pug304** [comment]
great! indeed this is working this way! Thanks for your explanation!

**Pull request merged**